lib/sensors/i2c_mux.hpp

```diff
@@ -0,0 +1,136 @@
+#pragma once
+
+#include "i2c_sensors.hpp"
+
+#include <array>
+#include <cstdint>
+#include <memory>
+#include <optional>
+
+#include <core/logger.hpp>
+#include <core/types.hpp>
+#include <io/i2c.hpp>
+
+namespace hyped::sensors {
+
+constexpr std::uint8_t kDefaultMuxAddress = 0x70;
+constexpr std::uint8_t kMaxNumChannels    = 8;
+// percentage of sensors that can be unusable before the mux is considered unusable
+constexpr core::Float kFailureThreshold = 0.25;  // TODOLater: finalize this value with Electronics
+
+/**
+ * @brief Mux for sensors using I2C
+ * @details Sensors include: Temperature (Wurth), Accelerometer (Wurth), Pressure (TODOLater: add
+ * name)
+ */
+template<class T, std::uint8_t N>
+class I2cMux {
+ public:
+  I2cMux(core::ILogger &logger,
+         std::shared_ptr<io::II2c> i2c,
+         const std::uint8_t mux_address,
+         std::array<std::unique_ptr<II2cMuxSensor<T>>, N> &sensors);
+  ~I2cMux();
+
+  std::optional<std::array<T, N>> readAllChannels();
+
+ private:
+  core::Result selectChannel(const std::uint8_t channel);
+  core::Result closeAllChannels();
+
+  core::ILogger &logger_;
+  std::shared_ptr<io::II2c> i2c_;
+  const std::uint8_t mux_address_;
+  const std::array<std::unique_ptr<II2cMuxSensor<T>>, N> sensors_;
+  const std::uint8_t max_num_unusable_sensors_;
+};
+
+template<typename T, std::uint8_t N>
+I2cMux<T, N>::I2cMux(core::ILogger &logger,
+                     std::shared_ptr<io::II2c> i2c,
+                     const std::uint8_t mux_address,
+                     std::array<std::unique_ptr<II2cMuxSensor<T>>, N> &sensors)
+    : logger_(logger),
+      i2c_(i2c),
+      mux_address_(mux_address),
+      sensors_(std::move(sensors)),
+      max_num_unusable_sensors_(static_cast<std::uint8_t>(kFailureThreshold * N))
+{
+  static_assert(N <= 8, "The I2c mux can only have up to 8 channels");
+}
+
+template<typename T, std::uint8_t N>
+I2cMux<T, N>::~I2cMux()
+{
+}
+
+template<typename T, std::uint8_t N>
+std::optional<std::array<T, N>> I2cMux<T, N>::readAllChannels()
+{
+  // Zero-initialize the array
+  std::array<T, N> mux_data{};
+  std::uint8_t num_unusable_sensors = 0;
+  for (std::uint8_t i = 0; i < N; ++i) {
+    const auto &sensor         = sensors_.at(i);
+    const std::uint8_t channel = sensor->getChannel();
+    // First ensure correct channel is selected
+    const auto channel_select_result = selectChannel(channel);
```

Comment:
idk if this is outside the scope of this pr but as I've read a lot of sensors code today, would making the `selectChannel` result optional and checking if `!channel_select_result` make more sense?

Reply:
no since we know every sensor on the mux must be initialised with a channel number (it cant ever be that a sensor doesn't have a channel number on the mux)